### PR TITLE
Support strictly blittable type marshalling

### LINKED ISSUE #69033: Support blittable type marshalling without `DisableRuntimeMarshalling`
Attempting to use `LibraryImport` in the aspnetcore repo (https://github.com/dotnet/aspnetcore/pull/41573) it was discovered that `DisableRuntimeMarshalling` will be needed for cases that don't seem to require that level of impact.

A small repo demonstrating the issue is below. I don't think this should require `DisableRuntimeMarshalling` if we can confirm the marshalling of `Foo` would be the same behavior (i.e., blittable).

```csharp
using System;
using System.Runtime.InteropServices;

namespace Demo
{
    internal struct Foo
    {
        public short A;
        public short B;
    }

    internal static partial class NativeExportsNE
    {
        public const string NativeExportsNE_Binary = "Microsoft.Interop.Tests." + nameof(NativeExportsNE);

        [LibraryImport(NativeExportsNE_Binary)]
        public static partial void PassFoo(Foo a);
    }

    internal static class Program
    {
        public static void Main(string[] args)
        {
            Foo a = default;
            NativeExportsNE.PassFoo(a);
        }
    }
}
```

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
This eases consumption of the `LibraryImport` source generator by not forcing the use of `DisableRuntimeMarshalling` for types where we can determine the type is blittable based on the following rules:

1) The type is "always blittable" (i.e., `int` or `double`, but not `char`).
1) A value type defined in the source project the current source generator is running on.
1) Is a type made up of types adhering to (1) and (2).

This will permit marshalling of the following:

```csharp
public struct S
{
    public short A;
    public short B;
}

[CustomTypeMarshaller(typeof(S))]
public struct Native
{
    private short a;
    private short b;
    public Native(S s) { ... }
    public S ToManaged() { ... }
}
```

Fixes https://github.com/dotnet/runtime/issues/69033

/cc @dotnet/interop-contrib

### CODE REVIEW

**Comment by jkotas** [comment]
The existing behavior is documented in https://github.com/dotnet/runtime/blob/main/docs/design/libraries/LibraryImportGenerator/StructMarshalling.md#design-2-disableruntimemarshallingattribute . Are we keeping this design doc up-to-date?

**Comment by AaronRobinsonMSFT** [comment]
> Are we keeping this design doc up-to-date?

Yep. This was simply to get feedback first. I thought had marked it "Draft". I do have one more test to add and will update doc once that passes.

File: docs/design/libraries/LibraryImportGenerator/StructMarshalling.md [hunk 1]
```diff
@@ -246,6 +246,22 @@ As an alternative design, we can use a different definition of "does not require
...
 
+Users are permitted to rely upon the built-in notion of blittable if the type to marshal adheres to the following constraints, termed "strictly blittable" in code:
+
+1) Is always blittable (for example, `int` or `double`, but not `char`).
+2) Is a value type defined in the source project the current source generator is running on.
+3) Is a type composed of types adhering to (1) and (2).
+
+The "strictly blittable" notion is being introduced to ease adoption. The impact of `DisableRuntimeMarshallingAttribute` is that some scenarios require effort that is much larger relative to the need to marshal an internal transfer type that has historically been blittable. For example, the following value type would require the consumer of the source generator to apply `DisableRuntimeMarshallingAttribute` if the above was not permitted.
```
Comment (elinor-fung) [inline docs/design/libraries/LibraryImportGenerator/StructMarshalling.md:255]:
```suggestion
The "strictly blittable" notion is being introduced to ease adoption. The impact of `DisableRuntimeMarshallingAttribute` can be quite large, impeding adoption in scenarios where the need is only to marshal an internal transfer type that has historically been blittable. For example, the following value type would require the consumer of the source generator to apply `DisableRuntimeMarshallingAttribute` if the above was not permitted.
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs [hunk 2]
```diff
@@ -13,42 +13,80 @@ namespace Microsoft.Interop
...
+                if (t.SpecialType is not SpecialType.None)
+                {
+                    return t.SpecialType.IsAlwaysBlittable();
+                }
+                else if (t.IsValueType)
                 {
-                    if (!IsConsideredBlittable(field.Type, seenTypes.Add(type)))
+                    // If the containing assembly for the type is backed by metadata (non-null),
+                    // then the type is not internal and therefore coming from a reference assembly
+                    // that we can not confirm is strictly blittable.
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs:67]: Can you add a test for this that checks we don't consider the referenced type blittable?

File: docs/design/libraries/LibraryImportGenerator/StructMarshalling.md (hunk 1 above)
Comment (jkotas) [inline docs/design/libraries/LibraryImportGenerator/StructMarshalling.md:255]: > some scenarios require effort that is much larger relative to the need to marshal an internal transfer type that has historically been blittable.

This contradicts statement made in the next paragraph "As we believe that users should be able to move over their assemblies to the new source-generated interop world as a whole assembly".

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs (hunk 2 above)
Comment (elinor-fung) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs:67]: Example test that creates a metadata reference from source and then a compilation that references it: https://github.com/dotnet/runtime/blob/d07dee0dabec8bc81d0d0ed042649c4c20c811d9/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexGeneratorParserTests.cs#L732-L750

File: docs/design/libraries/LibraryImportGenerator/StructMarshalling.md (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply docs/design/libraries/LibraryImportGenerator/StructMarshalling.md:255]: Great point all around. Will update.

**Review by elinor-fung — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
Failing leg is unrelated, Docker pull failure.

**Pull request merged**